### PR TITLE
[CalyxToFSM] Add FSM materialization pass

### PR DESCRIPTION
**Pull request by mortbopet**
In materializing a Calyx control FSM we:
* Gather up all of the `calyx.group`s referenced in the FSM
* Create an in- and output for each group, representing `done` (FSM input) and `go` (FSM output) signals
* Iterate over all states in the FSM and
  * replace `calyx.enable`s with an `fsm.output` bitvector asserting the corresponding `go` outputs
  * Extend the guards of all transitions from the state with a conjunction of the `done` signals of the enabled groups in the given state.
* The FSM is extended with a top-level go and done signal which are referenced in the `fsm_entry/fsm_exit` states, respectively.

The materialization does not consider how to materialize static groups. However, this should be fairly easy to implement/hook into when iterating over states and detecting `calyx.group`s with `static` attributes.

This pass was added under `CalyxToFSM` with the alternative being to make it a `Calyx` pass. Please let me know what you think is the most suitable.

Discussion in #3138.

### CODE REVIEW

File: lib/Conversion/CalyxToFSM/MaterializeFSM.cpp [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+namespace {
+
+// A set of group names. Use SetVector to ensure deterministic ordering.
+using GroupSet = SetVector<StringAttr>;
+
+struct MaterializeCalyxToFSMPass
+    : public MaterializeCalyxToFSMBase<MaterializeCalyxToFSMPass> {
+  void runOnOperation() override;
+
+  struct StateEnableInfo {
```
Comment (cgyurgyik) [inline lib/Conversion/CalyxToFSM/MaterializeFSM.cpp:36]: Can we add a comment here explaining what this is for?

File: lib/Conversion/CalyxToFSM/MaterializeFSM.cpp [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+    llvm::SmallVector<Value> outputOperands;
+    llvm::SmallVector<Value> doneGuards;
+  };
+
+  /// Returns an operand vector for the states which should be enabled for a
```
Comment (cgyurgyik) [inline lib/Conversion/CalyxToFSM/MaterializeFSM.cpp:41]: What is the consensus on using `///` versus `//` in LLVM?

File: lib/Conversion/CalyxToFSM/MaterializeFSM.cpp [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+  /// given state, as well as the state done signals that should be considered
+  /// for the state transition guard. We do this here to avoid
+  /// assignStateOutputOperands/assignStateTransitionGuard having to each walk
+  /// the set of output operations.
+  StateEnableInfo gatherStateEnableInfo(StateOp stateOp) {
+    StateEnableInfo info;
+    size_t portIndex = 0;
+    auto &enabledGroups = stateEnables[stateOp];
+    for (auto group : referencedGroups) {
```
Comment (cgyurgyik) [inline lib/Conversion/CalyxToFSM/MaterializeFSM.cpp:50]: Can we add a type here please?

File: lib/Conversion/CalyxToFSM/MaterializeFSM.cpp [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+      ++portIndex;
+    }
+    return info;
+  }
+
+  /// Assigns the 'fsm.output' operation of the provided 'state' to enabled the
+  /// set of provided groups. If 'topLevelDone' is set, also asserts the
+  /// top-level done signal.
+  void assignStateOutputOperands(StateOp stateOp,
+                                 llvm::SmallVector<Value> outputOperands,
```
Comment (cgyurgyik) [inline lib/Conversion/CalyxToFSM/MaterializeFSM.cpp:66]: If I'm understanding correctly, the intention here is to make a copy and then append a value. This seems a bit tricky and I wonder if something more explicit would be better.

File: lib/Conversion/CalyxToFSM/MaterializeFSM.cpp [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+                                 bool topLevelDone = false) {
+    assert(outputOperands.size() == machineOp.getNumArguments() - 1 &&
+           "Expected exactly one value for each uniquely referenced group in "
+           "this machine");
+    // outputOperands is expected to only have
+    outputOperands.push_back(topLevelDone ? c1 : c0);
+    auto outputOp = stateOp.output().getOps<fsm::OutputOp>();
+    assert(!outputOp.empty() &&
+           "Expected an fsm.output op inside the state output region");
+    (*outputOp.begin()).getOperation()->setOperands(outputOperands);
```
Comment (cgyurgyik) [inline lib/Conversion/CalyxToFSM/MaterializeFSM.cpp:76]: I imagine there is no .front( ) method here?

File: lib/Conversion/CalyxToFSM/MaterializeFSM.cpp [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+  GroupSet referencedGroups;
+
+  /// Maintain a relation between states and the groups which they enable.
+  DenseMap<fsm::StateOp, GroupSet> stateEnables;
+
+  /// A handle to the machine under transformation.
+  MachineOp machineOp;
+
+  /// Commonly used constants.
+  Value c1, c0;
```
Comment (cgyurgyik) [inline lib/Conversion/CalyxToFSM/MaterializeFSM.cpp:113]: Is this really cleaner than just having helper functions to create the constants, and letting another GVN pass do the cleanup? IMO, it just seems like extra things to reason about / a place where use-before-initialization may get ugly. Up to you.

File: test/Conversion/CalyxToFSM/materialize-fsm.mlir [hunk 7]
```diff
@@ -0,0 +1,139 @@
+// RUN: circt-opt -pass-pipeline='calyx.program(calyx.component(materialize-calyx-to-fsm))' %s | FileCheck %s
+
+// todo: non-deterministically fails due to state reordering.
```
Comment (cgyurgyik) [inline test/Conversion/CalyxToFSM/materialize-fsm.mlir:3]: Does this still apply?

File: lib/Conversion/CalyxToFSM/MaterializeFSM.cpp [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+    size_t portIndex = 0;
+    auto &enabledGroups = stateEnables[stateOp];
+    for (auto group : referencedGroups) {
+      if (enabledGroups.contains(group)) {
+        info.outputOperands.push_back(c1);
+        info.doneGuards.push_back(machineOp.getArgument(portIndex));
+      } else
+        info.outputOperands.push_back(c0);
+
+      ++portIndex;
```
Comment (cgyurgyik) [inline lib/Conversion/CalyxToFSM/MaterializeFSM.cpp:57]: (Suggestion) enumeration instead?

**Review by cgyurgyik — COMMENTED** [review]
Overall LGTM, I don't have a ton of experience with the FSM pass, so another set of eyes is a good idea.

File: lib/Conversion/CalyxToFSM/MaterializeFSM.cpp (hunk 4 above)
Comment (mortbopet) [reply lib/Conversion/CalyxToFSM/MaterializeFSM.cpp:66]: > make a copy

What are you referring to here? (copy of what?)


The intention is to in-place modify the operands of the `fsm.output` operation within the `output` region of the state based on the set of operands provided.

I agree that the implementation looks a bit strange given that `gatherStateEnableInfo` is where the operands are defined. Opted to do it this way (as described by the `gatherStateEnableInfo` comment) to avoid having to walk over the set of referenced groups for a state.

For the sake of readability, this is probably premature optimization - i think I'll go ahead sink `gatherStateEnableInfo`  into `assignStateOutputOperands/assignStateTransitionGuard` to make them more readable.

File: lib/Conversion/CalyxToFSM/MaterializeFSM.cpp (hunk 6 above)
Comment (mortbopet) [reply lib/Conversion/CalyxToFSM/MaterializeFSM.cpp:113]: This was an attempt to avoid having a`getOrCreateConstant`-like function (found in many other conversion functions) to avoid bloating the IR with unnecessary constant ops. I think it still makes sense to define these once and then reuse, but  use-before-initialization ugliness is a valid concern. I'll take another stab at it.

File: test/Conversion/CalyxToFSM/materialize-fsm.mlir (hunk 7 above)
Comment (mortbopet) [reply test/Conversion/CalyxToFSM/materialize-fsm.mlir:3]: Nope! will remove.

File: lib/Conversion/CalyxToFSM/MaterializeFSM.cpp (hunk 2 above)
Comment (mortbopet) [reply lib/Conversion/CalyxToFSM/MaterializeFSM.cpp:41]: [`///` where doxygen may pay the code a visit](https://llvm.org/docs/CodingStandards.html#id13).  
I usually just do `///` to preface functions and member variables, and `//` inside functions.

File: lib/Conversion/CalyxToFSM/MaterializeFSM.cpp (hunk 5 above)
Comment (mortbopet) [reply lib/Conversion/CalyxToFSM/MaterializeFSM.cpp:76]: correct.

File: lib/Conversion/CalyxToFSM/MaterializeFSM.cpp (hunk 4 above)
Comment (cgyurgyik) [reply lib/Conversion/CalyxToFSM/MaterializeFSM.cpp]: I mean outputOperands is passed by value and thus will make a copy

Comment (mortbopet) [reply lib/Conversion/CalyxToFSM/MaterializeFSM.cpp]: Ahh, it makes sense now. Nevertheless, this has been refactored in the latest commit.

File: lib/Conversion/CalyxToFSM/MaterializeFSM.cpp [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+        guardConjunction = guards.front();
+      else
+        guardConjunction = b.create<comb::AndOp>(transition.getLoc(), guards);
+      guardOp.setOperand(guardConjunction);
+    }
+  }
+
+  Value getOrCreateConstant(OpBuilder &b, APInt value) {
+    auto it = constants.find(value);
+    if (it != constants.end())
```
Comment (cgyurgyik) [inline lib/Conversion/CalyxToFSM/MaterializeFSM.cpp:90]:
```suggestion
    if (auto it = constants.find(value); it != constants.end())
```

Drive-by suggestion, since `it` isn't used in the rest of the function. This may be more of a personal thing though on the LLVM side of things.
